aea/aea.py

```diff
@@ -0,0 +1,290 @@
+
+"""This module contains the implementation of an Autonomous Economic Agent."""
+import base64
+import importlib.util
+import inspect
+import logging
+import os
+import re
+from abc import abstractmethod, ABC
+from typing import Optional, Dict
+
+from aea.protocols.base.abstract_handler import AbstractHandler
+
+from aea.agent import Agent
+from aea.mail.base import Envelope, ProtocolId
+from aea.protocols.base.protocol import Protocol
+from aea.protocols.default.message import DefaultMessage
+from aea.protocols.default.serialization import DefaultSerializer
+
+logger = logging.getLogger(__name__)
+
+SkillId = str
+
+
+class Registry(ABC):
+    """This class implements an abstract registry."""
+
+    @abstractmethod
+    def populate(self, directory: str) -> None:
+        """
+        Load into the registry as specified in the config and apply consistency checks.
+
+        :param directory: the filepath to the agent's resource directory.
+        :return: None
+        """
+
+    @abstractmethod
+    def teardown(self) -> None:
+        """
+        Teardown the registry.
+
+        :return: None
+        """
+
+
+class ProtocolRegistry(Registry):
+    """This class implements the handlers registry."""
+
+    def __init__(self) -> None:
+        """
+        Instantiate the registry.
+
+        :return: None
+        """
+        self._protocols = {}  # type: Dict[ProtocolId, Protocol]
+        self._handlers = {}  # type: Dict[ProtocolId, AbstractHandler]
+
+    def populate(self, directory: str) -> None:
+        """
+        Load the handlers as specified in the config and apply consistency checks.
+
+        :param directory: the filepath to the agent's resource directory.
+        :return: None
+        """
+
+        protocols_spec = importlib.util.find_spec(".".join([directory, "protocols"]))
+        protocols_packages = list(filter(lambda x: not x.startswith("__"), protocols_spec.loader.contents()))
+        logger.debug("Processing the following protocol packages: {}".format(protocols_packages))
+        for protocol_name in protocols_packages:
+            try:
+                self._add_protocol(directory, protocol_name)
+            except Exception:
+                logger.exception("Not able to add protocol {}.".format(protocol_name))
+
+    def fetch_protocol(self, envelope: Envelope) -> Optional[Protocol]:
+        """
+        Fetch the protocol for the envelope.
+
+        :pass envelope: the envelope
+        :return: the protocol id or None if the protocol is not registered
+        """
+        return self._protocols.get(envelope.protocol_id, None)
+
+    def fetch_handler(self, envelope: Envelope) -> Optional[AbstractHandler]:
+        """
+        Fetch the handler for the envelope.
+
+        :pass envelope: the envelope
+        :return: the protocol id or None if the protocol is not registered
+        """
+        return self._protocols.get(envelope.protocol_id, None)
```

Comment:
Not sure this is correct, it returns the same value as the previous function as it has the same return line...

Reply:
yeah, that's wrong. That should read the `.handler` property of the protocol object, if any.